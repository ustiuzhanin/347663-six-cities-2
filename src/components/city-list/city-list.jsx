import React, {useEffect} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {ActionsCreator} from "../../reducer";

const CityList = (props) => {
  useEffect(() => {
    const {offers, createListOfCities} = props;

    const listOfCities = [];
    offers.filter(({city}) => {
      if (!listOfCities.includes(city.name)) {
        listOfCities.push(city.name);
      }
    });

    createListOfCities(listOfCities);
  }, []);

  const handleCityLinkClick = (evt) => {
    const {activeCity, onCityLinkClick, resetOffersList} = props;

    onCityLinkClick(evt.currentTarget.id);
    if (activeCity !== evt.currentTarget.id) {
      resetOffersList();
    }
  };

  const cityList = (list, currentActiveCity) => {
    return list.map((city, i) => (
      <li className="locations__item" key={`${city}${i}`}>
        <a
          className={`locations__item-link tabs__item ${city ===
            currentActiveCity && `tabs__item--active`}`}
          href="#"
          id={city}
          onClick={handleCityLinkClick}
        >
          <span>{city}</span>
        </a>
      </li>
    ));
  };

  const {listOfCities, activeCity} = props;

  return cityList(listOfCities, activeCity);
};

CityList.propTypes = {
  offers: PropTypes.arrayOf(
      PropTypes.shape({
        city: PropTypes.shape({
          name: PropTypes.string.isRequired
        }).isRequired
      }).isRequired
  ).isRequired,
  listOfCities: PropTypes.array.isRequired,
  activeCity: PropTypes.string.isRequired,
  onCityLinkClick: PropTypes.func.isRequired,
  createListOfCities: PropTypes.func.isRequired,
  resetOffersList: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    activeCity: state.activeCity,
    listOfCities: state.listOfCities
  });

const mapDispatchToProps = (dispatch) => ({
  onCityLinkClick: (activeCityId) => {
    dispatch(ActionsCreator.changeCity(activeCityId));
  },
  createListOfCities: (listOfCities) => {
    dispatch(ActionsCreator.createListOfCities(listOfCities));
  },
  resetOffersList: () => {
    dispatch(ActionsCreator.resetOffersList());
  }
});

export {CityList};

export default connect(mapStateToProps, mapDispatchToProps)(CityList);
