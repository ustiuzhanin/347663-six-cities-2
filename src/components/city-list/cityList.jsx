import React from 'react';

export default function cityList() {
  return <div></div>;
}
