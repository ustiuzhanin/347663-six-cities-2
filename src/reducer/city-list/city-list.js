const initialState = {
  activeCity: 'Amsterdam',
  listOfCities: []
};

const ActionType = {
  CHANGE_CITY: 'CHANGE_CITY',
  CREATE_LIST_OF_CITIES: 'CREATE_LIST_OF_CITIES'
};

const ActionCreator = {
  changeCity: (city) => {
    return {
      type: ActionType.CHANGE_CITY,
      payload: city
    };
  },
  createListOfCities: (cities) => {
    return {
      type: ActionType.CREATE_LIST_OF_CITIES,
      payload: cities
    };
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return Object.assign({}, state, {
        activeCity: action.payload
      });
    case ActionType.CREATE_LIST_OF_CITIES:
      return Object.assign({}, state, {
        listOfCities: [...action.payload]
      });
  }

  return state;
};

export {ActionCreator, ActionType, reducer};
