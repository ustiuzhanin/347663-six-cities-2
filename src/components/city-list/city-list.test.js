import React from "react";
import { CityList } from "./city-list";
import ShallowRenderer from "react-test-renderer/shallow";

test(`CityList's snapshot`, () => {
  const renderer = new ShallowRenderer();

  const mockCity = {
    _id: "6005f36637751d35c8503011",
    name: `Dusseldorf`,
    location: { latitude: 51.225402, longitude: 6.776314, zoom: 13 },
  };

  renderer.render(
    <CityList
      activeCity={`Amsterdam`}
      listOfCities={[mockCity, mockCity, mockCity]}
      onCityLinkClick={jest.fn()}
      loadCityList={jest.fn()}
    />
  );
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});
