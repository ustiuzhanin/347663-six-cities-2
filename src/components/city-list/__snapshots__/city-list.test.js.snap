// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityList's snapshot 1`] = `
Array [
  <li
    className="locations__item"
  >
    <a
      className="locations__item-link tabs__item false"
      href="#"
      id="Dusseldorf"
      onClick={[Function]}
    >
      <span>
        Dusseldorf
      </span>
    </a>
  </li>,
  <li
    className="locations__item"
  >
    <a
      className="locations__item-link tabs__item false"
      href="#"
      id="Dusseldorf"
      onClick={[Function]}
    >
      <span>
        Dusseldorf
      </span>
    </a>
  </li>,
  <li
    className="locations__item"
  >
    <a
      className="locations__item-link tabs__item false"
      href="#"
      id="Dusseldorf"
      onClick={[Function]}
    >
      <span>
        Dusseldorf
      </span>
    </a>
  </li>,
]
`;
